src/micro.c

```diff
@@ -0,0 +1,680 @@
+}
+
+microError *
+microService_AddGroup(microGroup **new_group, microService *m, const char *prefix)
+{
+    if ((m == NULL) || (new_group == NULL) || (prefix == NULL))
+        return micro_ErrorInvalidArg;
+
+    *new_group = NATS_CALLOC(1, sizeof(microGroup) +
+                                    strlen(prefix) + 1); // "prefix.""
+    if (new_group == NULL)
+    {
+        return micro_ErrorOutOfMemory;
+    }
+
+    memcpy((*new_group)->prefix, prefix, strlen(prefix) + 1);
+    (*new_group)->m = m;
+    (*new_group)->next = m->groups;
+    m->groups = *new_group;
+
+    return NULL;
+}
+
+microError *
+microGroup_AddGroup(microGroup **new_group, microGroup *parent, const char *prefix)
+{
+    char *p;
+    int len;
+
+    if ((parent == NULL) || (new_group == NULL) || (prefix == NULL))
+        return micro_ErrorInvalidArg;
+
+    *new_group = NATS_CALLOC(1, sizeof(microGroup) +
+                                    strlen(parent->prefix) + 1 + // "parent_prefix."
+                                    strlen(prefix) + 1);         // "prefix\0"
+    if (new_group == NULL)
+    {
+        return micro_ErrorOutOfMemory;
+    }
+
+    p = (*new_group)->prefix;
+    len = strlen(parent->prefix);
+    memcpy(p, parent->prefix, len);
+    p[len] = '.';
+    p += len + 1;
+    memcpy(p, prefix, strlen(prefix) + 1);
+    (*new_group)->m = parent->m;
+    (*new_group)->next = parent->m->groups;
+    parent->m->groups = *new_group;
+
+    return NULL;
+}
+
+natsConnection *
+microService_GetConnection(microService *m)
+{
+    if (m == NULL)
+        return NULL;
+    return m->nc;
+}
+
+microError *
+microService_GetInfo(microServiceInfo **new_info, microService *m)
+{
+    microServiceInfo *info = NULL;
+    microEndpoint *ep = NULL;
+    int len = 0;
+
+    if ((new_info == NULL) || (m == NULL) || (m->service_mu == NULL))
+        return micro_ErrorInvalidArg;
+
+    info = NATS_CALLOC(1, sizeof(microServiceInfo));
+    if (info == NULL)
+        return micro_ErrorOutOfMemory;
+
+    info->Name = m->cfg->Name;
+    info->Version = m->cfg->Version;
+    info->Description = m->cfg->Description;
+    info->Id = m->id;
+    info->Type = MICRO_INFO_RESPONSE_TYPE;
+
+    // Overallocate subjects, will filter out internal ones.
+    info->Subjects = NATS_CALLOC(m->num_eps, sizeof(char *));
+    if (info->Subjects == NULL)
+    {
+        NATS_FREE(info);
+        return micro_ErrorOutOfMemory;
+    }
+
+    micro_lock_service(m);
+    for (ep = m->first_ep; ep != NULL; ep = ep->next)
+    {
+        if ((!ep->is_monitoring_endpoint) && (ep->subject != NULL))
+        {
+            info->Subjects[len] = ep->subject;
+            len++;
+        }
+    }
+    info->SubjectsLen = len;
+    micro_unlock_service(m);
+
+    *new_info = info;
+    return NULL;
+}
+
+void microServiceInfo_Destroy(microServiceInfo *info)
+{
+    if (info == NULL)
+        return;
+
+    // subjects themselves must not be freed, just the collection.
+    NATS_FREE(info->Subjects);
```

Comment:
Need to cast because Subjects is defined as `const`

Reply:
It's a `const char **`, which (perhaps again a #backfromgoland) I thought was a `(const char *) *`, which would imply the correct semantics. Am I wrong? (still, since a public API, better to dupe/free them separately).

Reply:
Yes, probably correct (if the compiler is not complaining, I think it's fine). As for freeing the underlying strings, I am not saying that you must, but if we don't do dupe/free, it has to be clearly stated that it is "owned" by the endpoint. Like for instance some getters in the library returns directly the pointer to the string and states that the user should not free this memory (in that case, the returned string is most likely as a `const char*` to emphasize that point).

Reply:
Yes, 2/5 in this case since performance is not an issue, and there's already a `Destroy` method needed - it's cleaner to dupe them. I was just in a hurry and/or lazy, will fix.

Reply:
Done (duplicate all strings now)

Reply:
But still, `info->Subjects` is a const and Windows warn about this free:
```
src\micro.c(580): warning C4090: 'function': different 'const' qualifiers
```

Reply:
@levb Just compiled on Windows and still get the warning. As suggested in previous comment, you may want to do the cast, just to silence Windows compiler.

Reply:
Sorry missed it, done now.